lib/msal-node/src/network/HttpClient.ts

```diff
@@ -267,7 +272,23 @@ const networkRequestViaHttps = <T>(
                     // do not destroy the request for the device code flow
                     networkResponse.body["error"] !== Constants.AUTHORIZATION_PENDING) {
                     request.destroy();
-                    reject(new Error(`HTTP status code ${statusCode}`));
+                    if (networkResponse.body["error"] === InteractionRequiredAuthErrorMessage.invalid_grant.code) {
```

Comment:
I don't think this the right approach. The network client should not have business logic for certain error codes, as that should be handled upstream. The key is that is when the network client receives a server error (e.g. token expired), it should not throw an error, and instead of should return a successful response (with the error information in the body of that response). The network clients should only throw an error when there is a network-level error, not when the server returns an error response.

Reply:
https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/0a0ca9b006f775309090fafa33fde6a9aae550ad/lib/msal-common/src/client/AuthorizationCodeClient.ts#L71-L83

Reply:
i.e. the logic to inspect the error code should likely be removed completely, and only reject when a network connection error is returned (which is what the browser fetch API does).

Reply:
And then the upstream code that is calling the network client should be repsonsible for inspecting the response and throwing an error, if needed.

Reply:
Thanks for the suggestions @jasonnutter , have updated the HttpClient to not reject based on error code, removing the custom descriptions for the errors;   the second part -"only reject when a network connection error is returned (which is what the browser fetch API does)" is already being handled by the [#3146](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/3146/files#diff-aa090b4d29f3b8c0c012e0788459864cdc3949a50939fb3da6b93958d473a4c7) as you mentioned as well.

Reply:
> I think that removing the if checks in HttpClient on lines [179](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/src/network/HttpClient.ts#L179) and [266](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/src/network/HttpClient.ts#L266) may be the solution. I had only added those checks because the old axios code (AxiosRequestConfig) had `validateStatus: () => true,`. According to [axios's documentation](https://github.com/axios/axios):
> 
> > // `validateStatus` defines whether to resolve or reject the promise for a given
> > // HTTP response status code. If `validateStatus` returns `true` (or is set to `null`
> > // or `undefined`), the promise will be resolved; otherwise, the promise will be
> > // rejected.
> > validateStatus: function (status) {
> > return status >= 200 && status < 300; // default
> > },
> 
> Per @jasonnutter's [comment](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/5175#discussion_r964094154), we are still (and apparently should only be) rejecting the promise for network-level errors on lines [190](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/src/network/HttpClient.ts#L190), [197](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/src/network/HttpClient.ts#L197) and [277](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/src/network/HttpClient.ts#L277).
> 
> The original axios code (before I rewrote the HttpClient) can be found [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/4744/files).

Robbie and I discussed this a while back and the issue comes from the added "validation" check that swallows all server errors and reports them as network errors. The confusion came from the axios callback option called "validateStatus" which we set as `validateStatus: () => true` (i.e. hardcoded it to return true so we could resolve with the server error and let the upstream code handle it). This was misinterpreted to mean `validateStatus: true` like we were "turning validation on" so when we replaced the HTTP module logic, a "validation if block" that reported all errors as HTTP errors was added.

Removing that if block should let the service errors bubble up to the code we already had to handle them.